File: ios/WhoCalled/Shared/WarmupLoader.swift
import CallKit
import Foundation
import OSLog

/// Preloads the embedded warm-up list (bundled `warmup_ios.json`) into the App
/// Group store on first launch, so iOS blocks/labels known FR spam immediately —
/// offline, before any network sync. No-op once the store is populated. After
/// this, ListService deltas keep the same single file up to date.
enum WarmupLoader {
  private static let logger = Logger(subsystem: "com.whocalled.app", category: "WarmupLoader")

  /// Returns true if the warm-up was just loaded (i.e. this is the first launch)
  /// so the caller can trigger an immediate first-time sync.
  @discardableResult
  static func loadIfNeeded() async -> Bool {
    guard SharedStore.readScoredNumbers().numbers.isEmpty else { return false }
    guard let url = Bundle.main.url(forResource: "warmup_ios", withExtension: "json") else {
      logger.notice("no warm-up asset bundled (warmup_ios.json)")
      return false
    }
    do {
      let data = try Data(contentsOf: url)
      let file = try JSONDecoder().decode(ScoredNumbersFile.self, from: data)
      guard !file.numbers.isEmpty else { return false }
      SharedStore.writeScoredNumbers(file)
      // Seed the sync cursor to the warm-up's serverTime so the next sync only
      // pulls deltas since the embedded baseline (not a full re-download).
      if let t = ISO8601DateFormatter().date(from: file.serverTime) {
        SharedStore.lastSyncAt = t
      }
      try await reloadExtension()
      logger.info("Warm-up loaded \(file.numbers.count) numbers")
      return true
    } catch {
      logger.error("warm-up load failed: \(error.localizedDescription)")
      return false
    }
  }

  private static func reloadExtension() async throws {
    try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
      CXCallDirectoryManager.sharedInstance.reloadExtension(
        withIdentifier: AppConstants.callDirectoryExtensionId
      ) { error in
        if let error { cont.resume(throwing: error) } else { cont.resume() }
      }
    }
  }
}

File: android/app/src/main/java/com/whocalled/android/data/WarmupLoader.kt
package com.whocalled.android.data

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import java.io.File

/**
 * Preloads the embedded warm-up database (assets/warmup.sqlite) into the single
 * Room cache on first launch, so the app blocks/labels known spam numbers
 * immediately — offline, before any network sync. No-op once the cache is
 * populated. After this, GET /lists deltas keep the same table up to date.
 */
object WarmupLoader {
    private const val TAG = "WarmupLoader"
    private const val ASSET = "warmup.sqlite"

    suspend fun loadIfNeeded(context: Context, db: WhoCalledDatabase) {
        if (db.scoredNumberDao().count() > 0) return

        val tmp = File(context.cacheDir, "warmup.sqlite")
        try {
            context.assets.open(ASSET).use { input ->
                tmp.outputStream().use { input.copyTo(it) }
            }
        } catch (e: Exception) {
            Log.w(TAG, "no warm-up asset bundled ($ASSET): ${e.message}")
            return
        }

        var loaded = 0
        try {
            SQLiteDatabase.openDatabase(tmp.path, null, SQLiteDatabase.OPEN_READONLY).use { src ->
                src.rawQuery(
                    "SELECT phone, status, spamScore, category, updatedAt FROM scored_numbers", null,
                ).use { c ->
                    val batch = ArrayList<ScoredNumberEntity>(2000)
                    while (c.moveToNext()) {
                        batch.add(
                            ScoredNumberEntity(
                                phone = c.getString(0),
                                status = c.getString(1),
                                spamScore = c.getInt(2),
                                category = c.getString(3),
                                updatedAt = c.getLong(4),
                            ),
                        )
                        if (batch.size >= 2000) {
                            db.scoredNumberDao().upsertAll(batch)
                            loaded += batch.size
                            batch.clear()
                        }
                    }
                    if (batch.isNotEmpty()) {
                        db.scoredNumberDao().upsertAll(batch)
                        loaded += batch.size
                    }
                }
            }
            Log.i(TAG, "Warm-up loaded $loaded numbers")
        } catch (e: Exception) {
            Log.e(TAG, "warm-up load failed: ${e.message}", e)
        } finally {
            tmp.delete()
        }
    }
}
